machine-tests/src/test/java/com/dylibso/chicory/testing/MachinesTest.java

```diff
@@ -0,0 +1,217 @@
+        moduleInstanceBuilder().withExternalValues(store.toExternalValues()).build();
+
+        // stderr?
+        assertEquals(expectedOutput, stderr.toString(UTF_8));
+
+        // and now runtime AOT
+        moduleInstanceBuilder()
+                .withMachineFactory(AotMachine::new)
+                .withExternalValues(store.toExternalValues())
+                .build();
+
+        assertEquals(expectedOutput + expectedOutput, stderr.toString(UTF_8));
+    }
+
+    // quickjs -> interpreter
+    // module -> pre-compiled aot / runtime aot
+    @Test
+    public void shouldRunQuickJsInterpreted() {
+        var stderr = new ByteArrayOutputStream();
+
+        var wasi = setupWasi(stderr);
+        // using the pre-compiled version of QuickJS
+        var quickjs =
+                quickJsInstanceBuilder()
+                        .withExternalValues(new ExternalValues(wasi.toHostFunctions()))
+                        .build();
+
+        var store = new Store().register("javy_quickjs_provider_v1", quickjs);
+
+        // the module is going to use the pre compiled aot
+        moduleInstanceBuilder()
+                .withMachineFactory(DynamicHelloJSMachineFactory::create)
+                .withExternalValues(store.toExternalValues())
+                .build();
+
+        // stderr?
+        assertEquals(expectedOutput, stderr.toString(UTF_8));
+
+        // and now runtime AOT
+        moduleInstanceBuilder()
+                .withMachineFactory(AotMachine::new)
+                .withExternalValues(store.toExternalValues())
+                .build();
+
+        assertEquals(expectedOutput + expectedOutput, stderr.toString(UTF_8));
+    }
+
+    // quickjs -> runtime aot
+    // module -> pre-compiled aot / interpreter
+    @Test
+    public void shouldRunQuickJsRuntimeAot() {
+        var stderr = new ByteArrayOutputStream();
+
+        var wasi = setupWasi(stderr);
+        // using the runtime aot version of QuickJS
+        var quickjs =
+                quickJsInstanceBuilder()
+                        .withMachineFactory(AotMachine::new)
+                        .withExternalValues(new ExternalValues(wasi.toHostFunctions()))
+                        .build();
+
+        var store = new Store().register("javy_quickjs_provider_v1", quickjs);
+
+        // the module is going to use the pre compiled aot
+        moduleInstanceBuilder()
+                .withMachineFactory(DynamicHelloJSMachineFactory::create)
+                .withExternalValues(store.toExternalValues())
+                .build();
+
+        // stderr?
+        assertEquals(expectedOutput, stderr.toString(UTF_8));
+
+        // and now the interpreter
+        moduleInstanceBuilder().withExternalValues(store.toExternalValues()).build();
+
+        assertEquals(expectedOutput + expectedOutput, stderr.toString(UTF_8));
+    }
+
+    @Test
+    public void shouldUseMachineCallOnlyForExport() throws Exception {
```

Comment:
Should we move this test to the AOT module? It's not dependent on precompilation

Reply:
good suggestion :+1: it's just a sanity check from the prev implementation of `call_indirect` at the end of the day.

Reply:
Turns out that is not dependent on precompilation, but it uses `wabt` and moving it to `aot` introduces a circular dependency.
Let's keep it as a follow up.